### PR TITLE
document Store and KVStore

### PR DESCRIPTION
**Pull request by n3wscott**
Trying to add notes about how and why to use State.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -328,6 +328,70 @@ spec:
 Will be processed first as a go template, then applied the Environment using the
 dynamic client.
 
+#### Communicating State
+
+There are times when the test author would rather pass state from a `Setup` or
+`Reqirement` of a feature without a complicated parameter passing or callbacks.
```
Comment (vaikas) [inline README.md:334]: Requirement

File: README.md [hunk 2]
```diff
@@ -328,6 +328,70 @@ spec:
...
+func someAssertion(ctx context.Context, t *testing.T) {
+    foo := state.GetOrFail(ctx, t, "foo")
+    // do something with foo.
+}
+```
+
+This shows `someAssertion` can be written in a way that does not require it to
+have some complicated callback logic wrapping it but still get custom parameters
+that are dependent on the specific test run. State is useful for passing
+generated names of know keys between `Setup` and `StepFn`s. It is intended to
```
Comment (vaikas) [inline README.md:387]: s/know keys/known keys/?

File: README.md [hunk 3]
```diff
@@ -328,6 +328,70 @@ spec:
...
+
+In use,
+
+```go
+func FancyFeature(brokerName string) *feature.Feature {
+	f := feature.NewFeatureNamed("MyNewFeature")
+
+	f.Setup("Set Foo", func(ctx context.Context, t *testing.T) {
+		state.SetOrFail(ctx, t, "foo", "bar-"+random)
+    })
```
Comment (lionelvillard) [inline README.md:370]: indent is off

Comment (n3wscott) [reply README.md]: thanks!

File: README.md [hunk 4]
```diff
@@ -328,6 +328,70 @@ spec:
...
+	})
+
+	f.Stable("An aspect of the feature").
+		Should("make some assertion", someAssertion)
+
+	return f
+}
+
+func someAssertion(ctx context.Context, t *testing.T) {
+    foo := state.GetOrFail(ctx, t, "foo")
```
Comment (lionelvillard) [inline README.md:379]: must be `GetStringOrFail`

File: README.md [hunk 5]
```diff
@@ -328,6 +328,70 @@ spec:
...
+// Direct Store access:
+store := state.FromContext(ctx) Store
+
+// Helpers to work with Store without needing fetch Store directly.
+
+aString := state.GetStringOrFail(ctx, t, key)
+
+state.GetOrFail(ctx, t, key, value)
+
+state.SetOrFail(ctx, t, key, value)
```
Comment (lionelvillard) [inline README.md:359]: Maybe mentioned here that under the cover the value is marshaled and only exported fields are stored.

File: README.md (hunk 4 above)
Comment (n3wscott) [reply README.md:379]: `GetStringOfFail` is a special case for known keys as strings

File: README.md [hunk 6]
```diff
@@ -328,6 +328,70 @@ spec:
...
+import knative.dev/reconciler-test/pkg/state
+
+// Direct Store access:
+store := state.FromContext(ctx) Store
+
+// Helpers to work with Store without needing fetch Store directly.
+
+aString := state.GetStringOrFail(ctx, t, key)
+
+state.GetOrFail(ctx, t, key, value)
```
Comment (n3wscott) [inline README.md:357]: value must be a pointer.

File: README.md (hunk 4 above)
Comment (n3wscott) [reply README.md:379]: AI: I need to followup to see if this is the right syntax here.

**Pull request merged**